src/storage/src/hummock/sstable/delete_range_aggregator.rs

```diff
@@ -0,0 +1,311 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::collections::{BTreeSet, BinaryHeap};
+use std::sync::Arc;
+
+use risingwave_hummock_sdk::key::user_key;
+use risingwave_hummock_sdk::key_range::KeyRange;
+use risingwave_hummock_sdk::HummockEpoch;
+
+use crate::hummock::DeleteRangeTombstone;
+
+pub struct SortedBoundary {
+    sequence: HummockEpoch,
+    user_key: Vec<u8>,
+}
+
+impl PartialEq<Self> for SortedBoundary {
+    fn eq(&self, other: &Self) -> bool {
+        self.user_key.eq(&other.user_key) && self.sequence == other.sequence
+    }
+}
+
+impl PartialOrd for SortedBoundary {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let ret = other
+            .user_key
+            .cmp(&self.user_key)
+            .then_with(|| other.sequence.cmp(&self.sequence));
+        Some(ret)
+    }
+}
+
+impl Eq for SortedBoundary {}
+
+impl Ord for SortedBoundary {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.user_key
+            .cmp(&other.user_key)
+            .then_with(|| other.sequence.cmp(&self.sequence))
+    }
+}
+
+pub struct DeleteRangeAggregator {
+    delete_tombstones: Vec<DeleteRangeTombstone>,
+    key_range: KeyRange,
+    watermark: u64,
+    gc_delete_keys: bool,
+}
+
+impl DeleteRangeAggregator {
+    pub fn new(key_range: KeyRange, watermark: u64, gc_delete_keys: bool) -> Self {
+        Self {
+            key_range,
+            delete_tombstones: vec![],
+            watermark,
+            gc_delete_keys,
+        }
+    }
+
+    pub fn add_tombstone(&mut self, data: Vec<DeleteRangeTombstone>) {
+        for mut tombstone in data {
+            if !self.key_range.left.is_empty() {
+                let split_start_user_key = user_key(&self.key_range.left);
+                if split_start_user_key.gt(tombstone.end_user_key.as_slice()) {
```

Comment:
`split_start_user_key` is inclusive while `tombstone.end_user_key` is exclusive. Should this be `ge` instead of `gt`?

Reply:
yes.

Reply:
fixed